### PR TITLE
Create parent directories if needed when untarring files+symlinks

### PR DESCRIPTION
**Pull request by babbageclunk**
A tarfile can be created without directory entries for all of the files included. Juju backup files have this, for example - they include home/ubuntu/.ssh/authorized_files without a directory entry for .ssh.

Files like this can be untarred fine with UntarFiles if they're extracted to an existing directory structure (so the .ssh directory already exists), but if they're extracted to an empty directory UntarFiles would fail. The tar command handles them fine though.

Change UntarFiles to create missing directories before extracting files or symlinks.

### CODE REVIEW

File: tar/tar_test.go [hunk 1]
```diff
@@ -298,6 +299,7 @@ func (t *TarSuite) TestUntarFilesHeadersIgnored(c *gc.C) {
 	c.Assert(err, gc.IsNil)
 
 	err = UntarFiles(&buf, t.cwd)
+	c.Assert(err, jc.ErrorIsNil)
```
Comment (anastasiamac) [inline tar/tar_test.go]: Niiice \o/

**Comment by babbageclunk** [comment]
$$merge$$

**Comment by babbageclunk** [comment]
!!build!!

**Comment by babbageclunk** [comment]
!!build!!

**Comment by babbageclunk** [comment]
$$merge$$

**Comment by babbageclunk** [comment]
Weird intermittent failure from utils/ssh:SSHGoCryptoCommandSuite.TestStrictHostChecksDifferentKeyTypes
```
12:37:42 FAIL: ssh_gocrypto_test.go:357: SSHGoCryptoCommandSuite.TestStrictHostChecksDifferentKeyTypes
12:37:42 
12:37:42 ssh_gocrypto_test.go:57:
12:37:42     c.Assert(err, jc.ErrorIsNil)
12:37:42 ... value *net.OpError = &net.OpError{Op:"read", Net:"tcp", Source:(*net.TCPAddr)(0xc0002338f0), Addr:(*net.TCPAddr)(0xc000233920), Err:(*os.SyscallError)(0xc0004e61a0)} ("read tcp 127.0.0.1:38431->127.0.0.1:44424: read: connection reset by peer")
12:37:42 
12:37:42 OOPS: 55 passed, 1 skipped, 1 FAILED
12:37:42 --- FAIL: TestPackage (8.05s)
12:37:42 FAIL
12:37:42 FAIL	github.com/juju/utils/ssh	8.062s
```
Haven't been able to reproduce with stress-race locally.

$$merge$$

**Comment by babbageclunk** [comment]
Uh, now various tests are failing because dependencies (zip, bzr, mongod) haven't been installed, even though they're installed in the Makefile. :/

$$merge$$

**Comment by babbageclunk** [comment]
Looks like that was a launchpad issue.

**Pull request merged**